### DISCUSSION TITLE
How to register an identity provider with different settings per tenant?

### DISCUSSION THREAD

**Discussion by xperiandri** [OrchardCMS/OrchardCore#10082] [Q&A] [answered]
I have a multi-tenant app that authenticates into Bitrix24 via OAuth2.
I have an OAuth2 provider for Bitrix24.
Each Bitrix24 tenant requires me to use its name and probably a different client id and a secret.
How to achieve that having that the identity provider is enabled for all tenants?

**Comment by maglini** [comment]
I have next code in Startup Tenant Module which working now.
"artsoft.bitrix24.ru" name of my tenant

**Is it correct way?** 
```

services.AddAuthentication(options =>
    {
        options.DefaultScheme = CookieAuthenticationDefaults.AuthenticationScheme;
    })
    .AddCookie(options =>
    {
        options.LoginPath = "/artsoft.bitrix24.ru/signin";
        options.LogoutPath = "/artsoft.bitrix24.ru/signout";
    })
    .AddBitrix24(options =>
    {
      //  options.CallbackPath = $"/artsoft.bitrix24.ru{options.CallbackPath}";
        Configuration.Bind("Bitrix24", options);
        options.SaveTokens = true;
        // TODO: Move method to a separate static class
        options.Events.OnCreatingTicket += async (OAuthCreatingTicketContext ctx) =>
        {
            var secretClient = ctx.HttpContext.RequestServices.GetRequiredService<SecretClient>();
            var log = ctx.HttpContext.RequestServices.GetRequiredService<ILoggerFactory>()
                .CreateLogger("KeyVault");
            var emailClaim = ctx.Principal?.GetClaim(ClaimTypes.Email);
            if (emailClaim is not null)
                try
                {
                    await secretClient.SetBitrix24RefreshTokenAsync(emailClaim, ctx.RefreshToken);
                }
                catch (Exception ex)
                {
                    log.LogError(ex, "Cannot save token for {email}", emailClaim);
                    throw;
                }

            // TODO: Remove old refresh tokens that expired
        };
    });
```

**Reply by maglini** [reply]
Another one issue. Is it correct to add **tenant name** in cookies?
```
  .AddCookie(options =>
  {
      options.LoginPath = "/artsoft.bitrix24.ru/signin";
      options.LogoutPath = "/artsoft.bitrix24.ru/signout";
  })
```

**Reply by xperiandri** [reply]
@deanmarcussen @sebastienro, the only crash we see now is an inability to unprotect a ticket on authentication.
Orchard Framework app redirects to the identity provider, successfully authenticates, redirects back, and here on cookies middleware crashes

**Reply by deanmarcussen** [reply]
You would need to share your data protection keys between tenants. The blob data protection provider should allow you to do that with templating

**Reply by maglini** [reply]
@deanmarcussen We added reference to _OrchardCore.DataProtection.Azure_ and updated appsettings.json: 
```json
  "OrchardCore": {
      "Default": {
          "State": "Running",
          "RequestUrlHost": null,
          "RequestUrlPrefix": null,
          "Features": ["OrchardCore.DataProtection.Azure"],
          "CustomTitle": "Default Tenant",
          "CustomSetting": "Custom setting for Default tenant"
      },

      "OrchardCore_DataProtection_Azure": {
          "ConnectionString": "UseDevelopmentStorage=true;",
          "ContainerName": "dataprotection",
          "BlobName": ""
      }
  }
```

Folder and DataProtectionKeys.xml created for tenant but we still have an exception **System.Security.Cryptography.CryptographicException: 'The key {80829b71-1241-4e6b-bd6e-8b0eadc3e75a} was not found in the key ring.'**
 What you mean by share your data protection keys between tenants ?

**Reply by xperiandri** [reply]
@deanmarcussen maybe we don't understand something? Do you mean that all keys from all tenants must be consolidated in the same blob folder? No folder per tenant as it appears now.

**Reply by ns8482e** [reply]
@xperiandri  @maglini  I guess for multi tenant  shouldn't `OrchardCore_DataProtection_Azure` to be under the tenant? e.g.
```json
 "OrchardCore": {
      "Default": {
          "State": "Running",
          "RequestUrlHost": null,
          "RequestUrlPrefix": null,
          "Features": ["OrchardCore.DataProtection.Azure"],
          "CustomTitle": "Default Tenant",
          "CustomSetting": "Custom setting for Default tenant",

         "OrchardCore_DataProtection_Azure": {
              "ConnectionString": "UseDevelopmentStorage=true;",
              "ContainerName": "dataprotection",
              "BlobName": ""
          }
      }
  }
```

**Reply by deanmarcussen** [reply]
It really depends what you’re trying to do, but the cookie is protected by data protection, so if you are sharing some identity stuff that isn’t modular used then you’ll have to figure out how to use the same keys for each tenant, or modular use the code you have written. 

The blob protection allows liquid templating to configure the folder names which could let you share the same folder. It’s in the docs

**Reply by xperiandri** [reply]
Just trying to make auth work for all tenants using Orchard Core Framework

**Comment by deanmarcussen** [comment]
Suggest a small repro repository there’s not enough information here to understand where you are going wrong

**Reply by maglini** [reply]
@deanmarcussen https://github.com/maglini/multi-tenancy-app

**Reply by xperiandri** [reply]
@deanmarcussen so what do you think about our sample? Were you managed to run it?

**Reply by xperiandri** [reply]
@sebastienros do you know a solution?

**Reply by maglini** [reply]
@ns8482e

**Reply by ns8482e** [reply]
I guess you don't need to have tenant prefix in cookie options may be just `~/` is enough
```
options.LoginPath = "~/signin";
options.LogoutPath = "~/signout";
```

**Comment by xperiandri** [comment, marked as answer]
So the issue was caused by moving from single-tenant to multi-tenant having sign-in cookie left from the single-tenant configuration. Hence that cookie could not be decoded.
Removing a cookie solved the issue.

**Reply by xperiandri** [reply]
@deanmarcussen this was a partial fix.

The issue was that the first tenant occupies OAuth authentication and then challenge redirect always happens to the first accessed tenant URL prefix.

So we removed all our stuff and tried to do the same what you do in Orchard Core CMS

**Web app**
``` C#
services.AddOrchardCore()
    .AddMvc()
    .WithTenants()
    .AddAzureTableShellsConfiguration()
    .ConfigureServices(s =>
    {
        s.AddRazorPages();
        s.AddSingleton<IShellHost, ShellHost>();
        s.AddSingleton<IShellDescriptorManagerEventHandler>(sp => sp.GetRequiredService<IShellHost>());
        s.AddSingleton<ITenantService, TenantService>();
        s.AddAuthentication(options => options.DefaultSignOutScheme = IdentityConstants.ApplicationScheme);
    })
    .Configure((app, routes, srv) =>
    {
        app.UseAuthorization();
        app.UseWelcomePage("/");
    });
```

**Module**
``` C#
services.ConfigureApplicationCookie(options =>
{
	var tenantName = shellSettings.Name;
	options.Cookie.Name = "orchauth_" + HttpUtility.UrlEncode(tenantName);
	options.LoginPath = $"/signin";
	options.LogoutPath = $"/signout";
});

services.TryAddEnumerable(
	ServiceDescriptor.Singleton<
		IPostConfigureOptions<
			Bitrix24AuthenticationOptions>,
			OAuthPostConfigureOptions<Bitrix24AuthenticationOptions, TenantBitrix24AuthenticationHandler>>());

services.TryAddSingleton<IPostConfigureOptions<Bitrix24AuthenticationOptions>, Bitrix24PostConfigureOptions>();
services.AddTransient<TenantBitrix24AuthenticationHandler>();
services.Configure<AuthenticationOptions>(options =>
{
	options.AddScheme<TenantBitrix24AuthenticationHandler>(
		Bitrix24AuthenticationDefaults.AuthenticationScheme,
		displayName: Bitrix24AuthenticationDefaults.DisplayName);
});
services.Configure<Bitrix24AuthenticationOptions>(Bitrix24AuthenticationDefaults.AuthenticationScheme, options => { });
```

But we get 404
No webpage was found for the web address: `https://localhost:44318/oauth/authorize/?client_id=&scope=&response_type=code&redirect_uri=https%3A%2F%2Flocalhost%3A44318%2Fartsoft.bitrix24.ru%2Fsignin-bitrix24&state=&response_mode=form_post`

We tried to prepend the URL with tenant name with no luck
`https://localhost:44318/artsoft.bitrix24.ru/oauth/authorize/?client_id=&scope=&response_type=code&redirect_uri=https%3A%2F%2Flocalhost%3A44318%2Fartsoft.bitrix24.ru%2Fsignin-bitrix24&state=&response_mode=form_post`

What do we miss? What handles `/oauth/authorize/` path? And is it tenant dependant?

**Reply by xperiandri** [reply]
@sebastienros, do you have any ideas?

**Reply by Skrypt** [reply]
I believe that the /oauth/authorize/ comes from the OpenId module which is tenant aware.

**Reply by xperiandri** [reply]
But we don't use it. We use Orchard Core Framework, not CMS. And it worked for the first called tenant when we register authentication as in regular ASP.NET Core app

**Reply by Skrypt** [reply]
To be honest I would need to try. But it should definitely work.
